File: fyz-src/frontend/src/api/graph.ts
import request from "./request";
import type { ApiResponse } from "./types";

export interface GraphNode {
  id: string;
  name: string;
  type: "Job" | "SkillArea" | "TechStack" | "TechPoint" | "KnowledgePoint" | "SourceDocument" | "GraphSnapshot";
  stack: string | null;
  level: string | null;
  description: string;
  importance: number | null;
  frequency: number | null;
  x: number | null;
  y: number | null;
  properties: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  relation: string;
  properties: Record<string, unknown>;
}

export interface GraphSubgraph {
  nodes: GraphNode[];
  edges: GraphEdge[];
  node_count: number;
  edge_count: number;
  snapshot_version: string | null;
  truncated: boolean;
  returned?: number;
  total_available?: number | null;
  next_cursor?: string | null;
  has_more?: boolean;
  query_scope?: string | null;
}

const EMPTY_GRAPH: GraphSubgraph = {
  nodes: [], edges: [], node_count: 0, edge_count: 0,
  snapshot_version: null, truncated: false, returned: 0, has_more: false,
  next_cursor: null,
};

export async function getOverview(params?: {
  cursor?: string;
  page_size?: number;
  max_layer?: 1 | 2 | 3;
  stack?: string;
  level?: string;
  keyword?: string;
}): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>("/graph/overview", { params });
  return res.data.data || EMPTY_GRAPH;
}

export async function getNodeNeighbors(
  nodeId: string,
  params?: { cursor?: string; page_size?: number; max_layer?: 1 | 2 | 3 | 4 | 5 },
): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>(
    `/graph/nodes/${encodeURIComponent(nodeId)}/neighbors`, { params },
  );
  return res.data.data || EMPTY_GRAPH;
}

export async function getPanorama(
  params?: {
    stack?: string;
    level?: string;
    node_type?: string;
    keyword?: string;
    limit?: number;
  }
): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>("/graph/panorama", { params });
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

export async function getNodeDetail(nodeId: string): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>(`/graph/nodes/${nodeId}`);
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

export async function expandNode(nodeId: string, depth: number = 2, limit: number = 300): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>("/graph/expand", {
    params: { node_id: nodeId, depth, limit }
  });
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

export async function searchNodes(q: string, types?: string, limit: number = 20): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>("/graph/search", {
    params: { q, types, limit }
  });
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

export async function getPath(fromId: string, toId: string, maxDepth: number = 6): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>("/graph/path", {
    params: { from_id: fromId, to_id: toId, max_depth: maxDepth }
  });
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

export async function getJobTree(jobId: number, depth: number = 5): Promise<GraphSubgraph> {
  const res = await request.get<ApiResponse<GraphSubgraph>>(`/graph/jobs/${jobId}/tree`, {
    params: { depth }
  });
  return res.data.data || { nodes: [], edges: [], node_count: 0, edge_count: 0, snapshot_version: null, truncated: false };
}

File: fyz-src/backend/app/schemas/graph.py
"""图谱同步、补全与查询 DTO。"""

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import BaseModel, Field

from app.core.agent_runtime import GraphEnrichmentOutput, KnowledgePointOutput, TechPointOutput
GraphNodeType = Literal[
    "Job", "SkillArea", "TechStack", "TechPoint",
    "KnowledgePoint", "SourceDocument", "GraphSnapshot",
]


class GraphSyncMode(str, Enum):
    full = "full"
    incremental = "incremental"


class GraphSyncRequest(BaseModel):
    mode: GraphSyncMode = GraphSyncMode.incremental
    enrich_top_skills: bool = True


class GraphNode(BaseModel):
    id: str
    name: str
    type: GraphNodeType
    stack: str | None = None
    level: str | None = None
    description: str = ""
    importance: float | None = None
    frequency: int | None = None
    x: float | None = None
    y: float | None = None
    properties: dict = Field(default_factory=dict)


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    relation: str
    properties: dict = Field(default_factory=dict)


class GraphSubgraph(BaseModel):
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    node_count: int = 0
    edge_count: int = 0
    snapshot_version: str | None = None
    truncated: bool = False
    returned: int = 0
    total_available: int | None = None
    next_cursor: str | None = None
    has_more: bool = False
    query_scope: str | None = None


class GraphSnapshotResponse(BaseModel):
    id: str
    version: str
    snapshot_type: str
    status: str
    node_count: int
    edge_count: int
    fact_count: int
    metadata: dict
    created_at: datetime
    completed_at: datetime | None


class GraphEnrichmentCandidateResponse(BaseModel):
    id: int
    snapshot_id: str
    skill_id: int
    skill_name: str
    candidate_data: dict
    evidence_source_ids: list[str]
    confidence: float
    machine_validation_status: str
    review_status: str
    publication_status: str
    review_note: str | None
    reviewed_at: datetime | None
    published_at: datetime | None
    lock_version: int
    agent_run_id: str | None
    created_at: datetime
    updated_at: datetime


class GraphEnrichmentCandidatePage(BaseModel):
    items: list[GraphEnrichmentCandidateResponse] = Field(default_factory=list)
    total: int
    page: int
    page_size: int
    machine_failed_pending_count: int = 0


class GraphEnrichmentBatchRejectResponse(BaseModel):
    rejected_count: int
    candidate_ids: list[int] = Field(default_factory=list)


class GraphEnrichmentReviewRequest(BaseModel):
    action: Literal["approve", "reject"]
    note: str | None = Field(default=None, max_length=500)
    lock_version: int = Field(ge=0)


class GraphEnrichmentPublishRequest(BaseModel):
    candidate_ids: list[int] = Field(default_factory=list, max_length=100)
